### PR TITLE
Committing UXUI-209 for Credit Card Number Input

### PR DESCRIPTION
**Pull request by manifestinteractive**
## Credit Card Number Input

New component for https://holidayextras.jira.com/browse/UXUI-209

Passes `npm test` and `npm run lint`

#### Attributes

`accepted` Array - Cards to accept ( 'amex', 'discover', 'mastercard' &amp; 'visa' )
`formatInput` Boolean - ( default false ) Whether to Format Input for easier reading
`showImages` Boolean - ( default false ) Whether to show Credit Card Images
`defaultValue` String - ( default empty ) Initial Value of Credit Card Number
`name` String - ( default 'credit-card-number' ) Name of Input
`id` String - ( default 'credit-card-number' ) ID for Input

## Example

```js
var cc = {
  images: true,
  format: true,
  number: '4485148248332663',
  label: 'Credit Card: ',
  accepted: ['visa','mastercard']
};

var example = (
  <UIToolkit.CreditCardNumber
    showImages={cc.images}
    formatInput={cc.format}
    defaultValue={cc.number}
    label={cc.label}
    accepted={cc.accepted} />
);

React.render(example, mountNode);
```

## Screenshot

screenshot

---
#### Developer Definition of Done/Quality Checklist (for PR author to complete BEFORE code review):
- [x] I have checked the [contributing document](../blob/master/CONTRIBUTING.md) and I'm happy for this to be reviewed.

#### Software Engineer or Developer review:
- [x] I’ve witnessed the work behaving as expected (this could be on the authors machine or screencast).
- [x] I’ve checked for coding anti-patterns.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve run all the tests locally and they pass.

#### Software Engineer or project guru review:
- [ ] I’ve witnessed the work behaving as expected (this could be on the authors machine or screencast).
- [ ] I’ve checked for coding anti-patterns.
- [ ] I’ve checked for appropriate test coverage.
- [ ] I’ve run all the tests locally and they pass.

### CODE REVIEW

**Comment by stuartstobie** [comment]
Looking at this now

File: src/components/credit-card-number/code/templates/creditCardNumberTemplate.jsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+module.exports = function(component) {
+
+  var label, visa, mastercard, amex, discover, images;
+
+  if (component.props.label){
+    label = ( <label className="input-group-label credit-card-number-label" htmlFor={component.props.id}>{component.props.label}</label> );
+  }
+
+  if(component.props.showImages)
+  {
```
Comment (stuartstobie) [inline src/components/credit-card-number/code/templates/creditCardNumberTemplate.jsx]: Keep the { on the same lie as the if, as you have done in the previous if statement. Same applies below.

**Comment by cameronviner** [comment]
Taking DEV

**Comment by cameronviner** [comment]
Conflicts on this PR, will need to merge in master

File: docs/assets/style.css [hunk 2]
```diff
@@ -130,3 +130,41 @@ article {
 .toolkit-example button {
   margin:.25em;
 }
+
+/* Credit Card Number */
+.ui-component-credit-card-number .credit-card-images
```
Comment (cameronviner) [inline docs/assets/style.css]: Could we have the opening bracket on the same line as the css selector?

**Comment by cameronviner** [comment]
There doesn't seem to be a limit on the amount of characters I can enter on the credit card number field, should there be a limit on the number of characters the user can enter in the field?

**Comment by manifestinteractive** [comment]
Addressed comments and merge conflict in Nav.jsx file.  Master has been merged into this PR.

**Comment by manifestinteractive** [comment]
__Addressed the following:__

1.  Merged in master and addressed merge conflicts
2.  Updated Unit Tests format to new standard 
3. verified npm test && npm run lint pass

File: src/components/credit-card-number/code/templates/creditCardNumberTemplate.jsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    images = ( <div className="credit-card-images">{visa} {mastercard} {amex} {discover}</div> );
+  }
+
+  return (
+    <div className="component-credit-card-number">
+      {images}
+      <div className="credit-card-number">
+        {label}
+        <input
```
Comment (robhuzzey) [inline src/components/credit-card-number/code/templates/creditCardNumberTemplate.jsx]: Can we make use of the pattern attribute in HTML5 to help out here?: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-pattern

File: src/components/credit-card-number/code/views/creditCardNumberView.jsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      cardNumber: formattedCard,
+      cardType: cardInfo.type
+    };
+
+    this.setState(card);
+
+    return card;
+  },
+
+  getCard: function(cardNumber) {
```
Comment (robhuzzey) [inline src/components/credit-card-number/code/views/creditCardNumberView.jsx]: May be overkill, but there's a way to get card type in a handy little library: https://github.com/Shortbreaks/iinChecker

File: docs/src/Components.jsx [hunk 5]
```diff
@@ -81,6 +81,29 @@ var Components = React.createClass({
           </ul>
         </article>
 
+        <article>
+          <h3 id="credit_card_number">Credit Card Number</h3>
+          <p>Masked Credit Card Number input</p>
```
Comment (robhuzzey) [inline docs/src/Components.jsx]: The number isn't "Masked" as I expected... is this a correct description? I would have expected to see it mask as I type (like a password field) based on this description.

**Comment by robhuzzey** [comment]
Looks good to me... few comments, nothing major.

**Comment by cameronviner** [comment]
The cards images seem to have got bigger: https://cloudup.com/cI4LVG1tHCo and entering a card number no longer greys out/colours in the appropriate card.

Also if you try to amend any number apart form the last two it jumps to the end of the input. To recreate this enter the VISA number and then try and delete 85.

**Comment by manifestinteractive** [comment]
This is ready for review. Forgot to update CSS class names after removing the ui- prefix ( as requested ).  Added selection range to address input cursor jumping to end if editing text within input.  Also merged in latest master.

File: docs/src/Components.jsx (hunk 5 above)
Comment (jodiedoubleday) [inline docs/src/Components.jsx]: Tons of usability studies say don't mask, same thing with passwords now

**Comment by jodiedoubleday** [comment]
There is still a slight issue with the credit card field.
When you type in a credit card number (rather than copy and paste) the formatting goes AWOL
Screencast here - http://quick.as/wmwyfg25q

I hope the screencast makes sense

**Comment by manifestinteractive** [comment]
Swapped out custom Validation for third-party Payment library.  Also added error message but made sure it was a property that can be passed in, or set to empty, if desired.

File: src/components/credit-card-number/code/views/creditCardNumberView.jsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      formatInput: false,
+      id: 'credit-card-number',
+      name: 'credit-card-number',
+      placeHolder: '• • • •   • • • •   • • • •   • • • •',
+      showImages: false,
+      errorMessage: 'Invalid Card Number'
+    };
+  },
+
+  handleChange: function(elm) {
```
Comment (robhuzzey) [inline src/components/credit-card-number/code/views/creditCardNumberView.jsx]: Being pedantic here but the argument ```elm``` here makes me thing you are passing an element into the handle change method when in fact it's an event (convention is that it's ```e```)

**Comment by robhuzzey** [comment]
Only one comment about a variable name but apart from that... +1 from me.

**Comment by cameronviner** [comment]
Looks good to me, :+1:

**Comment by manifestinteractive** [comment]
@cameronviner @robhuzzey @jodiedoubleday Can someone press that fancy green "Merge pull request" button?

File: src/components/credit-card-number/code/views/creditCardNumberView.jsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  getDefaultProps: function() {
+    return {
+      accepted: ['amex', 'discover', 'mastercard', 'visa'],
+      formatInput: false,
+      id: 'credit-card-number',
+      name: 'credit-card-number',
+      placeHolder: '• • • •   • • • •   • • • •   • • • •',
+      showImages: false,
+      errorMessage: 'Invalid Card Number'
+    };
```
Comment (lukehansell-hx) [inline src/components/credit-card-number/code/views/creditCardNumberView.jsx]: no default is supplied for defaultValue and not providing one on the element declaration displays the invalid error for some reason.

File: src/components/credit-card-number/scripts/build-dev.sh [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Comment (lukehansell-hx) [inline src/components/credit-card-number/scripts/build-dev.sh]: can be removed

**Comment by manifestinteractive** [comment]
@lukehansell-hx I updated this component using the other credit card date feedback as well.  Has better unit tests and addresses the other comments listed.

**Pull request closed without merging**